.dispatchPerAllele_CollapsedVCF <- function(FUN, x, singleAltOnly) {
    als <- VariantAnnotation::alt(x)
    flat <- BiocGenerics::unlist(als, use.names=FALSE)
    res <- FUN(rep(VariantAnnotation::ref(x), S4Vectors::elementNROWS(VariantAnnotation::alt(x))), flat)
    lst <- BiocGenerics::relist(res, als)
    if (singleAltOnly)
        all(lst) & S4Vectors::elementNROWS(lst) == 1
    else
        any(lst)
}
.dispatchPerAllele_ExpandedVCF <- function(FUN, x) {
    als <- VariantAnnotation::alt(x)
    flat <- BiocGenerics::unlist(als, use.names=FALSE)
    res <- FUN(rep(VariantAnnotation::ref(x), S4Vectors::elementNROWS(VariantAnnotation::alt(x))), flat)
    res
}

#' Determines whether the variant is a symbolic allele
#'
#' @export
setGeneric("isSymbolic", signature="x",
           function(x, ...)
               standardGeneric("isSymbolic")
)
setMethod("isSymbolic", "CollapsedVCF",
          function(x, ..., singleAltOnly=TRUE)
              .dispatchPerAllele_CollapsedVCF(.isSymbolic, x, singleAltOnly)
)
setMethod("isSymbolic", "ExpandedVCF",
          function(x, ...)
              .dispatchPerAllele_ExpandedVCF(.isSymbolic, x)
)
.isSymbolic <- function(r, a) {
    result <- grepl("<", a, fixed=TRUE) |
        grepl("[", a, fixed=TRUE) |
        grepl("]", a, fixed=TRUE) |
    	grepl(".", a, fixed=TRUE)
    return(result)
}

#' Determines whether the variant is a structural variant
#'
#' @export
setGeneric("isStructural", signature="x",
           function(x, ...)
               standardGeneric("isStructural")
)
setMethod("isStructural", "CollapsedVCF",
          function(x, ..., singleAltOnly=TRUE)
              .dispatchPerAllele_CollapsedVCF(.isStructural, x, singleAltOnly)
)
setMethod("isStructural", "ExpandedVCF",
          function(x, ...)
              .dispatchPerAllele_ExpandedVCF(.isStructural, x)
)
.isStructural <- function(ref, alt) {
	lengthDiff <- S4Vectors::elementNROWS(ref) != IRanges::nchar(alt)
	if (is(alt, "DNAStringSet")) {
		# don't break if there are no symbolic alleles in the VCF
		return(lengthDiff)
	}
    return(as.logical(
    	# exclude no-call sites
    	!is.na(alt) & alt != "<NON_REF>" &
        (lengthDiff | .isSymbolic(ref, alt))))
}


#' Returns the structural variant length of the first allele
#'
#' @param vcf VCF object
#'
.svLen <- function(vcf) {
	assertthat::assert_that(.hasSingleAllelePerRecord(vcf))
    r <- VariantAnnotation::ref(vcf)
    a <- elementExtract(VariantAnnotation::alt(vcf))
    result <- ifelse(!isStructural(vcf), 0,
		elementExtract(VariantAnnotation::info(vcf)$SVLEN) %na%
		(elementExtract(VariantAnnotation::info(vcf)$END) - start(SummarizedExperiment::rowRanges(vcf))) %na%
		(ifelse(isSymbolic(vcf), NA_integer_, IRanges::nchar(a) - IRanges::nchar(r))))
    return(result)
}

.hasSingleAllelePerRecord <- function(vcf) {
	assertthat::assert_that(is(vcf, "VCF"))
    all(S4Vectors::elementNROWS(VariantAnnotation::alt(vcf)) == 1)
}
setMethod("isStructural", "VCF",
		  function(x, ...)
		  	.dispatchPerAllele_ExpandedVCF(.isStructural, x)
)


#' Extracts the structural variants as a BreakendGRanges
#'
#' @details
#' Structural variants are converted to breakend notation.
#'
#' Due to ambiguities in the VCF specifications, structural variants
#' with multiple alt alleles are not supported.
#'
#' If HOMLEN or HOMSEQ is defined without CIPOS, it is assumed that
#' the variant position is left aligned.
#'
#' @export
setGeneric("breakpointRanges", signature="x",
		   function(x, ...)
		   		standardGeneric("breakpointRanges")
)
setMethod("breakpointRanges", "VCF",
		  function(x, ...)
		  	.breakpointRanges(x, ...)
)

#' @param vcf VCF object
#' @param nominalPosition Determines whether to call the variant at the
#'    nominal VCF position, or to call the confidence interval (incorporating
#'    any homology present).
#' @param prefix variant name prefix to assign to unnamed variants
#' @param suffix suffix to append
#' @param info_columns VCF INFO columns to include in the GRanges object
.breakpointRanges <- function(vcf, nominalPosition=FALSE, placeholderName="svrecord", suffix="_bp", info_columns=NULL, unpartneredBreakends=FALSE) {
	vcf <- vcf[isStructural(vcf),]
	assertthat::assert_that(.hasSingleAllelePerRecord(vcf))
	# VariantAnnotation bug: SV row names are not unique
	# ensure names are defined
	if (any(duplicated(row.names(vcf)))) {
		warning("Found ", sum(duplicated(row.names(vcf))), " duplicate row names (duplicates renamed).")
	}
	if (is.null(row.names(vcf))) {
		row.names(vcf) <- paste0(placeholderName, seq_along(vcf), row.names(vcf))
	} else if (any(is.na(row.names(vcf)) | duplicated(row.names(vcf)))) {
		row.names(vcf) <- ifelse(is.na(row.names(vcf)) | duplicated(row.names(vcf)), paste0(placeholderName, seq_along(vcf)), row.names(vcf))
	}
	assertthat::assert_that(!is.null(row.names(vcf)))
	assertthat::assert_that(assertthat::noNA(row.names(vcf)))
	assertthat::assert_that(!any(duplicated(row.names(vcf))))
	gr <- SummarizedExperiment::rowRanges(vcf)
	gr$REF <- as.character(VariantAnnotation::ref(vcf))
	gr$ALT <- as.character(elementExtract(VariantAnnotation::alt(vcf), 1))
	gr$vcfId <- names(vcf)
	gr$partner <- rep(NA_character_, length(gr))
	gr$svtype <- elementExtract(VariantAnnotation::info(vcf)$SVTYPE) %na%
		# hack ensure that [,2] exists even for zero record vcfs
		(stringr::str_match(c("HACK", gr$ALT), "<(.*)>")[,2][-1]) %na%
		rep(NA_character_, length(gr))
	# use the root type
	gr$svtype <- stringr::str_extract(gr$svtype, "^[^:]+")
	gr$svLen <- .svLen(vcf)
	gr$insSeq <- rep(NA_character_, length(gr))
	gr$insLen <- rep(0, length(gr))
	gr$cistartoffset <- rep(0, length(gr))
	gr$ciwidth <- rep(0, length(gr))

	for (col in info_columns) {
		mcols(gr)[[col]] <- VariantAnnotation::info(vcf)[[col]]
	}
	if (!is.null(VariantAnnotation::info(vcf)$HOMSEQ)) {
		seq <- elementExtract(VariantAnnotation::info(vcf)$HOMSEQ, 1)
		gr$ciwidth <- ifelse(is.na(seq), gr$ciwidth, nchar(seq))
	}
	if (!is.null(VariantAnnotation::info(vcf)$HOMLEN)) {
		gr$ciwidth <- elementExtract(VariantAnnotation::info(vcf)$HOMLEN, 1) %na% gr$ciwidth
	}
	# have not yet factored in imprecise variant calling into ciwidth - just microhomology
	gr$HOMLEN <- gr$ciwidth

	if (!is.null(VariantAnnotation::info(vcf)$CIPOS)) {
		.expectMetadataInfo(vcf, "CIPOS", 2, header.Type.Integer)
		cistartoffset <- elementExtract(VariantAnnotation::info(vcf)$CIPOS, 1)
		ciendoffset <- elementExtract(VariantAnnotation::info(vcf)$CIPOS, 2)
		ciwidth <- ciendoffset - cistartoffset
		gr$cistartoffset <- cistartoffset %na% gr$cistartoffset
		gr$ciwidth <- ciwidth %na% gr$ciwidth
	}
	gr$processed <- rep(FALSE, length(gr))
	outgr <- gr[FALSE,]

	# Another workaround for single breakend variants turning into the empty string
	rows <- !gr$processed & !isSymbolic(vcf) & stringr::str_length(gr$ALT) > 0
	if (any(rows)) {
		cgr <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (!unpartneredBreakends) {
			commonPrefixLength <- pairwiseLCPrefix(cgr$REF, cgr$ALT, ignore.case=TRUE)
			cgr$svLen <- nchar(cgr$ALT) - nchar(cgr$REF)
			cgr$insSeq <- XVector::subseq(cgr$ALT, start=commonPrefixLength + 1)
			cgr$insLen <- nchar(cgr$insSeq)
			start(cgr) <- start(cgr) - 1 + commonPrefixLength
			width(cgr) <- 1
			strand(cgr) <- "+"
			mategr <- cgr
			strand(mategr) <- "-"
			ranges(mategr) <- IRanges::IRanges(start=BiocGenerics::start(cgr) + nchar(cgr$REF) - commonPrefixLength + 1, width=1)

			names(mategr) <- paste0(names(cgr), suffix, 2)
			names(cgr) <- paste0(names(cgr), suffix, 1)
			cgr$partner <- names(mategr)
			mategr$partner <- names(cgr)
			outgr <- c(outgr, cgr, mategr)
		}
		cgr <- NULL
		mategr <- NULL
	}
	rows <- !gr$processed & !is.na(gr$svtype) & gr$svtype %in% c("DEL", "INS", "DUP", "RPL")
	if (any(rows)) {
		cgr <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (!unpartneredBreakends) {
			cvcf <- vcf[rows,]
			#assertthat::assert_that(!any(cgr$svtype == "DEL" & cgr$svLen > 0))
			#assertthat::assert_that(!any(cgr$svtype == "INS" & cgr$svLen < 0))
			dup <- cgr$svtype == "DUP"
			del <- cgr$svtype == "DEL"
			ins <- cgr$svtype == "INS"

			BiocGenerics::strand(cgr) <- "+"
			BiocGenerics::width(cgr) <- 1
			cgr$insLen <- ifelse(ins, abs(cgr$svLen), 0)
			if (!is.null(VariantAnnotation::info(cvcf)$NTLEN)) {
				#pindel RPL
				cgr$insLen <- elementExtract(VariantAnnotation::info(cvcf)$NTLEN) %na% cgr$insLen
			}
			mategr <- cgr
			BiocGenerics::strand(mategr) <- "-"
			# use end, then fall back to calculating from length
			end <- elementExtract(VariantAnnotation::info(cvcf)$END, 1) %na% (BiocGenerics::start(cgr) + ifelse(ins, 0, abs(cgr$svLen)))
			if (any(is.na(end))) {
				stop(paste("Variant of undefined length: ", paste(names(cgr)[is.na(end),], collapse=", ")))
			}
			ranges(mategr) <- IRanges::IRanges(start=BiocGenerics::end + ifelse(dup, 0, 1), width=1)

			cistartoffset <- elementExtract(VariantAnnotation::info(cvcf)$CIEND, 1)
			ciendoffset <- elementExtract(VariantAnnotation::info(cvcf)$CIEND, 2)
			ciwidth <- ciendoffset - cistartoffset
			mategr$cistartoffset <- cistartoffset %na% mategr$cistartoffset
			mategr$ciwidth <- ciwidth %na% mategr$ciwidth


			BiocGenerics::strand(cgr)[dup] <- "-"
			BiocGenerics::strand(mategr)[dup] <- "+"

			names(mategr) <- paste0(names(cgr), suffix, 2)
			names(cgr) <- paste0(names(cgr), suffix, 1)
			cgr$partner <- names(mategr)
			mategr$partner <- names(cgr)
			outgr <- c(outgr, cgr, mategr)
		}
		cgr <- NULL
		mategr <- NULL
	}
	rows <- !gr$processed & !is.na(gr$svtype) & gr$svtype %in% c("INV")
	if (any(rows)) {
		cgr1 <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (!unpartneredBreakends) {
			width(cgr1) <- 1
			end <- elementExtract(VariantAnnotation::info(vcf)$END[rows], 1) %na% (BiocGenerics::start(cgr1) + abs(cgr1$svLen) - 1)
			if (any(is.na(end))) {
				stop(paste("Variant of undefined length: ", paste(names(cgr1)[is.na(end),], collapse=", ")))
			}
			hasPlusBreakend <- rep(TRUE, length(cgr1))
			hasMinusBreakend <- rep(TRUE, length(cgr1))
			if (!is.null(VariantAnnotation::info(vcf)$INV3)) {
				hasMinusBreakend <- !VariantAnnotation::info(vcf)$INV3[rows]
			}
			if (!is.null(VariantAnnotation::info(vcf)$INV5)) {
				hasPlusBreakend <- !VariantAnnotation::info(vcf)$INV5[rows]
			}

			cgr2 <- cgr1
			cistartoffset <- elementExtract(VariantAnnotation::info(vcf)$CIEND[rows], 1)
			ciendoffset <- elementExtract(VariantAnnotation::info(vcf)$CIEND[rows], 2)
			ciwidth <- ciendoffset - cistartoffset
			cgr2$cistartoffset <- cistartoffset %na% cgr2$cistartoffset
			cgr2$ciwidth <- ciwidth %na% cgr2$ciwidth
			cgr3 <- cgr1
			cgr4 <- cgr2

			ranges(cgr2) <- IRanges::IRanges(start=BiocGenerics::end + 1, width=1)
			ranges(cgr3) <- IRanges::IRanges(start=BiocGenerics::start(cgr1) - 1, width=1)
			ranges(cgr4) <- IRanges::IRanges(start=BiocGenerics::end, width=1)
			strand(cgr1) <- "-"
			strand(cgr2) <- "-"
			strand(cgr3) <- "+"
			strand(cgr4) <- "+"

			names(cgr4) <- paste0(names(cgr1), suffix, 4)
			names(cgr3) <- paste0(names(cgr1), suffix, 3)
			names(cgr2) <- paste0(names(cgr1), suffix, 2)
			names(cgr1) <- paste0(names(cgr1), suffix, 1)
			cgr1$partner <- names(cgr2)
			cgr2$partner <- names(cgr1)
			cgr3$partner <- names(cgr4)
			cgr4$partner <- names(cgr3)

			outgr <- c(outgr, cgr1[hasMinusBreakend], cgr2[hasMinusBreakend], cgr3[hasPlusBreakend], cgr4[hasPlusBreakend])
		}
		cgr1 <- NULL
		cgr2 <- NULL
		cgr3 <- NULL
		cgr4 <- NULL
	}
	rows <- !gr$processed & !is.na(gr$svtype) & gr$svtype %in% c("BND") & (stringr::str_detect(gr$ALT, stringr::fixed("[")) | stringr::str_detect(gr$ALT, stringr::fixed("]")))
	if (any(rows)) {
		cgr <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (!unpartneredBreakends) {
			cvcf <- vcf[rows,]

			bndMatches <- stringr::str_match(cgr$ALT, "(.*)(\\[|])(.*)(\\[|])(.*)")
			preBases <- bndMatches[,2]
			bracket <- bndMatches[,3]
			remoteLocation <- bndMatches[,4]
			postBases <- bndMatches[,6]
			strand(cgr) <- ifelse(preBases == "", "-", "+")

			cgr$partner <- NA_character_
			if (!is.null(VariantAnnotation::info(cvcf)$PARID)) {
				cgr$partner <- elementExtract(VariantAnnotation::info(cvcf)$PARID, 1)
			}
			if (!is.null(VariantAnnotation::info(cvcf)$MATEID) & any(is.na(cgr$partner))) {
				multimates <- S4Vectors::elementNROWS(VariantAnnotation::info(cvcf)$MATEID) > 1 & is.na(cgr$partner)
				cgr$partner <- ifelse(is.na(cgr$partner), elementExtract(info(cvcf)$MATEID, 1), cgr$partner)
				if (any(multimates)) {
					warning(paste("Ignoring additional mate breakends for variants", names(cgr)[multimates]))
				}
			}
			reflen <- S4Vectors::elementNROWS(cgr$REF)
			cgr$insSeq <- paste0(stringr::str_sub(preBases, reflen + 1), stringr::str_sub(postBases, end=-(reflen + 1)))
			cgr$insLen <- nchar(cgr$insSeq)

			toRemove <- is.na(cgr$partner) | !(cgr$partner %in% names(cgr))
			if (any(toRemove)) {
				warning(paste("Removing", sum(toRemove), "unpaired breakend variants", paste0(names(cgr)[toRemove], collapse=", ")))
				cgr <- cgr[!toRemove,]
			}
			mategr <- cgr[cgr$partner,]
			cgr$svLen <- ifelse(GenomeInfoDb::seqnames(cgr)==GenomeInfoDb::seqnames(mategr), abs(BiocGenerics::start(cgr) - BiocGenerics::start(mategr)) - 1, NA_integer_)
			# make deletion-like events have a -ve svLen
			cgr$svLen <- ifelse(BiocGenerics::strand(cgr) != BiocGenerics::strand(mategr) &
					((BiocGenerics::start(cgr) < BiocGenerics::start(mategr) & BiocGenerics::strand(cgr) == "+") |
					 (BiocGenerics::start(cgr) > BiocGenerics::start(mategr) & BiocGenerics::strand(cgr) == "-")),
				-cgr$svLen, cgr$svLen)
			cgr$svLen <- cgr$svLen + cgr$insLen
			outgr <- c(outgr, cgr)
		}
		cgr <- NULL
		mategr <- NULL
	}
	# breakends that are not in breakpoint notation should be in breakend notation
	rows <- !gr$processed & !is.na(gr$svtype) & gr$svtype %in% c("BND")
	if (any(rows)) {
		cgr <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (unpartneredBreakends) {
			cvcf <- vcf[rows,]
			BiocGenerics::strand(cgr) <- ifelse(cgr$ALT == "", "*", ifelse(stringr::str_sub(cgr$ALT, 1, 1) == ".", "-", "+"))
			# trim anchoring base and breakend symbol
			cgr$insSeq <- stringr::str_sub(cgr$ALT, 2, stringr::str_length(cgr$ALT) - 1)
			cgr$insLen <- stringr::str_length(cgr$insSeq)
			cgr$partner <- NA_character_
			cgr$svLen <- NA_integer_
			outgr <- c(outgr, cgr)
		}
		cgr <- NULL
	}
	# TODO: Does delly write two records for a full INV?
	# DELLY TRA https://groups.google.com/forum/#!msg/delly-users/6Mq2juBraRY/BjmMrBh3GAAJ
	rows <- !gr$processed & !is.na(gr$svtype) & gr$svtype %in% c("TRA")
	if (any(rows)) {
		cgr <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (!unpartneredBreakends) {
			cvcf <- vcf[rows,]

			if (is.null(VariantAnnotation::info(cvcf)$CHR2) || any(is.na(VariantAnnotation::info(cvcf)$CHR2))) {
				stop(paste("Delly variants missing CHR2:", paste(names(cgr)[is.na(VariantAnnotation::info(cvcf)$CHR2)], collapse=", ")))
			}
			if (is.null(VariantAnnotation::info(cvcf)$CT) || any(is.na(VariantAnnotation::info(cvcf)$CT))) {
				stop(paste("Delly variants missing CT:", paste(names(cgr)[is.na(VariantAnnotation::info(cvcf)$CT)], collapse=", ")))
			}
			cgr$insLen <- VariantAnnotation::info(cvcf)$INSLEN %na% 0 # Delly no longer writes INSLEN to all TRA records
			BiocGenerics::width(cgr) <- 1
			mategr <- cgr
			# Hack so we can add new seqlevels if required
			GenomeInfoDb::seqlevels(mategr) <- unique(c(GenomeInfoDb::seqlevels(mategr), VariantAnnotation::info(cvcf)$CHR2))
			GenomeInfoDb::seqnames(mategr)[seq(1, length(mategr))] <- VariantAnnotation::info(cvcf)$CHR2
			IRanges::ranges(mategr) <- IRanges::IRanges(start=VariantAnnotation::info(cvcf)$END, width=1)
			BiocGenerics::strand(cgr) <- ifelse(VariantAnnotation::info(cvcf)$CT %in% c("3to3", "3to5"), "+", "-")
			BiocGenerics::strand(mategr) <- ifelse(VariantAnnotation::info(cvcf)$CT %in% c("3to3", "5to3"), "+", "-")

			mcistartoffset <- elementExtract(VariantAnnotation::info(cvcf)$CIEND, 1) %na% 0
			mciendoffset <- elementExtract(VariantAnnotation::info(cvcf)$CIEND, 2) %na% 0
			mciwidth <- mciendoffset - mcistartoffset
			mategr$cistartoffset <- mcistartoffset
			mategr$ciwidth <- mciwidth

			names(mategr) <- paste0(names(cgr), suffix, 2)
			names(cgr) <- paste0(names(cgr), suffix, 1)
			cgr$partner <- names(mategr)
			mategr$partner <- names(cgr)
			outgr <- c(outgr, cgr, mategr)
		}
		cgr <- NULL
		mategr <- NULL
	}
	# TIGRA CTX
	rows <- !gr$processed & !is.na(gr$svtype) & gr$svtype %in% c("CTX")
	if (any(rows)) {
		# TIGRA CTX call
		cgr <- gr[rows,]
		gr$processed[rows] <- TRUE
		if (!unpartneredBreakends) {
			cvcf <- vcf[rows,]

			if (is.null(VariantAnnotation::info(cvcf)$CHR2) || any(is.na(VariantAnnotation::info(cvcf)$CHR2))) {
				stop(paste("TIGRA variants missing CHR2:", paste(names(cgr)[is.na(VariantAnnotation::info(cvcf)$CHR2)], collapse=", ")))
			}
			BiocGenerics::width(cgr) <- 1
			mategr <- cgr
			# Hack so we can add new seqlevels if required
			GenomeInfoDb::seqlevels(mategr) <- unique(c(GenomeInfoDb::seqlevels(mategr), VariantAnnotation::info(cvcf)$CHR2))
			GenomeInfoDb::seqnames(mategr)[seq(1, length(mategr))] <- VariantAnnotation::info(cvcf)$CHR2
			IRanges::ranges(mategr) <- IRanges::IRanges(start=VariantAnnotation::info(cvcf)$END, width=1)
			# no direction information is reported
			BiocGenerics::strand(cgr) <- "*"
			BiocGenerics::strand(mategr) <- "*"

			mcistartoffset <- elementExtract(VariantAnnotation::info(cvcf)$CIEND, 1) %na% 0
			mciendoffset <- elementExtract(VariantAnnotation::info(cvcf)$CIEND, 2) %na% 0
			mciwidth <- mciendoffset - mcistartoffset
			mategr$cistartoffset <- mcistartoffset
			mategr$ciwidth <- mciwidth

			names(mategr) <- paste0(names(cgr), suffix, 2)
			names(cgr) <- paste0(names(cgr), suffix, 1)
			cgr$partner <- names(mategr)
			mategr$partner <- names(cgr)
			outgr <- c(outgr, cgr, mategr)
		}
		cgr <- NULL
		mategr <- NULL
	}
	if (!all(gr$processed)) {
		stop(paste("Unrecognised format for variants", paste(names(gr)[!gr$processed], collapse=", ")))
	}
	# incorporate microhomology and confidence intervals
	if (!nominalPosition) {
		ranges(outgr) <- IRanges::IRanges(start=start(outgr) + outgr$cistartoffset, width=outgr$ciwidth + 1, names=names(outgr))
	}
	outgr$processed <- NULL
	outgr$cistartoffset <- NULL
	outgr$ciwidth <- NULL
	if (!unpartneredBreakends) {
		partnerpartnerisself <- partner(outgr)$partner == names(outgr)
		if (!all(partnerpartnerisself)) {
			warning("Multiple breakends partners for a single breakend found (Ignoring all except first). StructuralVariantAnnotation does not yet support promiscuous breakpoints. Please raise a github enhancement request at https://github.com/d-cameron/StructuralVariantAnnotation/issues if this functionality is desired.")
			outgr <- outgr[partnerpartnerisself,]
		}
		# sanity check that all breakpoints partners actually exist
		haspartner <- outgr$partner %in% names(outgr)
		if (!all(haspartner)) {
			stop(paste("Sanity check failure: unpaired breakends ", paste(names(gr)[!haspartner], collapse=", ")))
		}
	} else {
		outgr$partner <- NULL
	}
	return(outgr)
}
#' Extracts unpartnered breakend structural variants as a GRanges
#'
#' See Section 5.4.9 of https://samtools.github.io/hts-specs/VCFv4.3.pdf for details.
#'
#' @export
setGeneric("breakendRanges", signature="x",
		   function(x, ...)
		   	standardGeneric("breakendRanges")
)
setMethod("breakendRanges", "VCF",
		  function(x, ...)
		  	.breakpointRanges(x, unpartneredBreakends=TRUE, ...)
)

#' @param vcf VCF object
#' @param nominalPosition Determines whether to call the variant at the
#'    nominal VCF position, or to call the confidence interval (incorporating
#'    any homology present).
#' @param prefix variant name prefix to assign to unnamed variants
#' @param suffix suffix to append
#' @param info_columns VCF INFO columns to include in the GRanges object
.breakendRanges <- function(vcf, nominalPosition=FALSE, placeholderName="svrecord", suffix="_bp", info_columns=NULL) {
}

.hasMetadataInfo <- function(vcf, field) {
	return(field %in% row.names(VariantAnnotation::info(header(vcf))))
}
.expectMetadataInfo <- function(vcf, field, number, type) {
	assertthat::assert_that(.hasMetadataInfo(vcf, field))
	row <- VariantAnnotation::info(header(vcf))[field,]
	assertthat::assert_that(number == row$Number)
	assertthat::assert_that(type == row$Type)
}

